### PR TITLE
Check conditions, duplicates and empty config files

### PR DESCRIPTION
**Pull request by dnestoro**
## What does this PR do?

Checks content of config files, whether they:

- have duplicated entries
- have entries with missing condition
- have `java.lang.Object` as typeReachable value
- are empty
- contain entries that don't belong to library

In order to perform this check, run `./gradlew checkConfigFiles --coordinates groupID:artifactID:versionID`

This check should be executed in github actions automatically.

### CODE REVIEW

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/CoordinateUtils.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package org.graalvm.internal.tck;
+
+public abstract class CoordinateUtils {
```
Comment (fniephaus) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/CoordinateUtils.java:3]:
```suggestion
public class CoordinateUtils {
```

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/CoordinateUtils.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        return template
+                .replace("$group$", coordinates.group())
+                .replace("$sanitizedGroup$", coordinates.sanitizedGroup())
+                .replace("$artifact$", coordinates.artifact())
+                .replace("$sanitizedArtifact$", coordinates.sanitizedArtifact())
+                .replace("$capitalizedSanitizedArtifact$", coordinates.capitalizedSanitizedArtifact())
+                .replace("$version$", coordinates.version());
+    }
+
+
```
Comment (fniephaus) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/CoordinateUtils.java:15]:
```suggestion

```

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    void run() throws IllegalArgumentException {
+        Coordinates coordinates = Coordinates.parse(this.coordinates);
+
+        Path coordinatesMetadataRoot = getProject().file(CoordinateUtils.replace("metadata/$group$/$artifact$/$version$", coordinates)).toPath();
+        if (!coordinatesMetadataRoot.toFile().exists()) {
+            throw new IllegalArgumentException("ERROR: Cannot find metadata directory for given coordinates: " + this.coordinates);
+        }
+
+        boolean containsErrors = false;
+        List<File> filesInMetadata = getConfigFilesForMetadataDir(coordinatesMetadataRoot);
```
Comment (fniephaus) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:45]:
```suggestion
        File coordinatesMetadataRoot = getProject().file(CoordinateUtils.replace("metadata/$group$/$artifact$/$version$", coordinates));
        if (!coordinatesMetadataRoot.exists()) {
            throw new IllegalArgumentException("ERROR: Cannot find metadata directory for given coordinates: " + this.coordinates);
        }

        boolean containsErrors = false;
        List<File> filesInMetadata = getConfigFilesForMetadataDir(coordinatesMetadataRoot);
```

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        }
+
+        if (containsErrors) {
+            throw new IllegalStateException("Errors above found for: " + this.coordinates);
+        }
+    }
+
+    private List<File> getConfigFilesForMetadataDir(Path root) throws RuntimeException {
+        List<File> files = new ArrayList<>();
+        File [] content = root.toFile().listFiles();
```
Comment (fniephaus) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:75]:
```suggestion
    private List<File> getConfigFilesForMetadataDir(File root) throws RuntimeException {
        List<File> files = new ArrayList<>();
        File [] content = root.listFiles();
```

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+            if (entryName == null) {
+                entryName = entry.toString();
+            }
+
+            containsErrors = true;
+            System.out.println("ERROR: In file " + file.toURI() + " entry: " + entryName + " contains illegal typeReachable value. Field" +
+                    " typeReachable cannot be any of the following values: " + ILLEGAL_TYPE_VALUES);
+        }
+
+        return containsErrors;
```
Comment (fniephaus) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:269]:
```suggestion
        String typeReachable = (String) condition.get("typeReachable");
        if (ILLEGAL_TYPE_VALUES.stream().anyMatch(type -> type.equalsIgnoreCase(typeReachable))) {
            // get name of entry that misses typeReachable. If name cannot be determinate, write whole entry
            String entryName = (String) entry.get("name");
            if (entryName == null) {
                entryName = entry.toString();
            }
            System.out.println("ERROR: In file " + file.toURI() + " entry: " + entryName + " contains illegal typeReachable value. Field" +
                    " typeReachable cannot be any of the following values: " + ILLEGAL_TYPE_VALUES);
            return true;
        }

        return false;
```

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+    private boolean listNullOrEmpty(List list) {
+        if (list == null) {
+            return true;
+        }
+
+        return list.size() == 0;
```
Comment (fniephaus) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:277]:
```suggestion
        return list == null || list.size() == 0;
```

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            System.out.println("ERROR: empty reflect-config detected: " + file.toURI());
+            return true;
+        }
+
+        containsErrors |= containsDuplicatedEntries(entries, file);
+        for (var entry : entries) {
+            containsErrors |= checkTypeReachable(entry, file);
+        }
+
+        return containsErrors;
```
Comment (fniephaus) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:120]:
```suggestion
        List<Map<String, Object>> entries = getConfigEntries(file);

        if (entries.size() == 0) {
            System.out.println("ERROR: empty reflect-config detected: " + file.toURI());
            return true;
        }

        boolean containsErrors = containsDuplicatedEntries(entries, file);
        for (var entry : entries) {
            containsErrors |= checkTypeReachable(entry, file);
        }

        return containsErrors;
```

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        }
+
+        return containsDuplicatedEntries(entries, file);
+    }
+
+    private boolean containsDuplicatedEntries(List<Map<String, Object>> entries, File file) {
+        Map<Map<String, Object>, Integer> duplicates = new HashMap<>();
+        entries.forEach(entry -> {
+            duplicates.merge(entry, 1, Integer::sum);
+        });
```
Comment (fniephaus) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:227]:
```suggestion
        entries.forEach(entry -> duplicates.merge(entry, 1, Integer::sum));
```

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+            System.out.println("ERROR: empty proxy-config detected: " + file.toURI());
+            return true;
+        }
+
+        containsErrors |= containsDuplicatedEntries(entries, file);
+        for (var entry : entries) {
+            containsErrors |= checkTypeReachable(entry, file);
+        }
+
+        return containsErrors;
```
Comment (fniephaus) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:209]:
```suggestion
        List<Map<String, Object>> entries = getConfigEntries(file);

        if (entries.size() == 0) {
            System.out.println("ERROR: empty proxy-config detected: " + file.toURI());
            return true;
        }

        boolean containsErrors = containsDuplicatedEntries(entries, file);
        for (var entry : entries) {
            containsErrors |= checkTypeReachable(entry, file);
        }

        return containsErrors;
```

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+        File index = getProject().file("metadata/index.json");
+        getIndexFile().set(index);
+
+        this.allowedPackages = getAllowedPackages();
+    }
+
+
+    @TaskAction
+    void run() throws IllegalArgumentException, FileNotFoundException {
+        if (coordinates.group().equalsIgnoreCase("org.example") || coordinates.group().equalsIgnoreCase("samples")) {
```
Comment (gradinac) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:61]: Perhaps we could also move these two libraries into a "library names that we don't check" list

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+        }
+
+        if (containsErrors) {
+            throw new IllegalStateException("Errors above found for: " + this.coordinates);
+        }
+    }
+
+    private List<File> getConfigFilesForMetadataDir(File root) throws RuntimeException {
+        List<File> files = new ArrayList<>();
+        File [] content = root.listFiles();
```
Comment (gradinac) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:101]: Extra space between `File` and `[]`

File: tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+        if (entries.isEmpty()) {
+            System.out.println("ERROR: empty serialization-config detected: " + file.toURI());
+            return true;
+        }
+
+        boolean containsErrors = containsDuplicatedEntries(entries, file);
+        for (var entry : entries) {
+            containsErrors |= checkTypeReachable(entry, file);
+            containsErrors |= containsEntriesNotFromLibrary(entry, file);
+        }
```
Comment (gradinac) [inline tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ConfigFilesChecker.java:202]: This block could be a separate function which we then use here and in the two places where we check the new serialization config

**Review by gradinac — APPROVED** [review]
Looks good :) Merging this in the current form with more fixes coming up in a follow up PR

**Pull request merged**